Add a "create" state, to create a file without spurious 'changed'

mscherer opened the pull request:
This permit to be used with this pattern :

```
- file:
    path: "/srv/git/{{ item.name }}/git-daemon-export-ok"
    state: "{% if item.private %}absent{% else %}creates{% endif %}"
  with_items: git_repositories
```

And git_repositories being a array of item with a member "private".
Using copy/template would requires to used 2 actions instead
of 1, with duplication.

bcoca commented:
This should do what you want w/o adding extra code
- copy: content="\n" dest=/srv/git/{{item.name}}/git-daemon-export-ok force=no

mscherer commented:
No, because this doesn't remove the file if item.private is set. As I say in the initial comment, I can do 2 actions, 1 to remove, 1 to create the file, but this PR permit to have 1 action that can take a different parameter instead of 2.

In turn this permit to reduce duplication of the information, see http://new.theopensourceway.org/cgit/ansible/commit/?id=530fae32e00ead64d695d6e1db8525a78946a620 for the commit that implement the 2 action, vs the previous one, where this was 1 single action, but with 'touch'. Both work, but 'touch' make spurious changes, and this PR aim at giving a way to eliminate this change without having to resort using a totally different module.

florin-andrei-tp commented:
The file module not being able to do anything if the file does not already exist seems like such an artificial limitation. Anything that would remove that limit is welcome.

FlorinAndrei commented:
The file module not being able to do anything if the file does not exist already seems like such an artificial limitation. Anything that would remove that block is a welcome change.

mscherer commented:
Redid the patch. I splitted in several chunk so it is easy to review, as I was not really able to rediff cleanly and confidently. So there is 2 separates commits, and a 3rd one to reindent. But most of the code should be in the 2 first ones.

gregdek commented:
@mscherer Thanks for your pull request. This module is critical to Ansible's functionality, so it's owned by the Core team, and it can take a while for patches to be reviewed. Thanks for your patience here.

gregdek commented:
Pushing to next release. Thanks @mscherer for your continued patience with this one.

gregdek commented:
Thanks @mscherer for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

mscherer committed:
Add a "create" state, to create a file without spurious 'changed'

This permit to use this pattern:

    - file:
        path: "/srv/git/{{ item.name }}/git-daemon-export-ok"
        state: "{% if item.private %}absent{% else %}create{% endif %}"
      with_items: git_repositories

And git_repositories being a array of item with a member "private".
Using copy/template would requires to used 2 actions instead
of 1, with duplication of information and more complexity.

gregdek commented:
Thanks @mscherer for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

[This message brought to you by your friendly Ansibull-bot.]

robinro commented on files/file.py:
```diff
@@ -56,16 +56,18 @@
         If C(touch) (new in 1.4), an empty file will be created if the C(path) does not
         exist, while an existing file or directory will receive updated file access and
         modification times (similar to the way `touch` works from the command line).
+        If C(creates) (new in 2.1), an empty file will be created, but a existing file will
```
typo: **an** existing instead of **a** existing

mscherer replied:
fixed, thanks

robinro commented:
Do I see correctly that if you do creates=yes on an existing directory it does nothing and gives changed=False?
That would be confusing if one expects an empty file. I'm not convinced this is better than having 2 separate tasks.

mscherer commented:
I am not sure to understand, there is no "creates=yes", but "state=create". I didn't tested that case, but I would expect it to react like "state=touch" on a directory (which would then be equally confusing because that's a behavior that already exist)

robinro commented:
I put two thoughts in one sentence, one is the one you answered and the other is that `state=creates` should be `state=create`, to be in line with `touch`. This also reduces the possible confusion with `creates=/filename` in the command module.

mscherer commented:
Mhh, good point. Or state=created would sound better, but I am not a native speaker, so maybe it doesn't sound great. Let's remove the 's' in the mean time, it seems more correct.

mscherer commented:
So another use case:
https://gitlab.com/osas/ansible-role-mailman3/blob/master/tasks/logs.yml#L11

and 
https://gitlab.com/osas/ansible-role-mailman3/blob/master/tasks/logs.yml#L17

While it can be argued that both operation belong to the rpm (and that's likely how we would fix it), this is a example where "create" would do the trick.

gregdek commented:
Thanks @mscherer for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

mscherer commented:
Gosh, my own other PR is the one breaking this one :(

mscherer commented:
ready_for_review

robinro commented on files/file.py:
```diff
@@ -51,16 +51,18 @@
         do not exist, since 1.7 they will be created with the supplied permissions.
-        If C(file), the file will NOT be created if it does not exist, see the M(copy)
-        or M(template) module if you want that behavior.  If C(link), the symbolic
+        If C(file), the file will NOT be created if it does not exist, see C(creates)
```
typo: should be 'create' not 'creates'

gregdek commented:
Thanks @mscherer for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

duck-rh commented:
Any news? Noone seemed to object on the code and the usefulness of the feature was already discussed, but nothing happened since May.

gregdek commented:
Thanks @mscherer for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@mscherer A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@mscherer Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

mscherer commented:
ready_for_review

mscherer commented:
I also have a commit for test once this is merged

gregdek commented:
Thanks @mscherer for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]
